### PR TITLE
Offer card form properties and method on BTDropInViewController

### PR DESCRIPTION
**Pull request by burnto**
This change allows for optional customization and pre-population of the card form within the drop-in view controller.

@skunkworks @intelliot @mickeyreiss any feedback / suggestions?

### CODE REVIEW

File: Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m [hunk 1]
```diff
@@ -51,6 +64,16 @@ - (void)tappedToShowDropIn {
...
     dropIn.shouldHideCallToAction = NO;
 
+    if (self.prepopulateDataSwitch.on) {
+        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
+        [dateFormatter setDateFormat:@"MM/yyyy"];
+        [dropIn.cardForm setExpirationDate:[dateFormatter dateFromString: @"12/2018"]];
+        dropIn.cardForm.number = @"4111111111111111";
+        dropIn.cardForm.cvv = @"123";
+        dropIn.cardForm.postalCode = @"12345";
+        dropIn.cardForm.optionalFields = BTUICardFormOptionalFieldsAll;
```
Comment (mickeyreiss) [inline Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m]: Nice! This brings us a big step closer towards easy integration with card.io in Drop in.

**Comment by mickeyreiss** [comment]
I'm :+1: on the change overall. Thanks for taking care of this!

Two suggestions:

1) I think we already have a `card.io` demo in our demo app. Maybe we should switch that over to Drop in, since that is an often requested mash-up.

2) In my opinion, we'll end up with a clearer and more understandable public API if we wrap the relevant features of `BTCardFormView` with thin delegator methods. In practice this would mean adding a few methods to `BTDropInViewController`: `setNumber:`, `setCvv:`, `setPostalCode:`, `setOptionalFields:`, setAlphanumericPostalCode:`.

**Comment by burnto** [comment]
@mickeyreiss I've incorporated your feedback on number 2 above.

Instead of the bitmasking options, I elected to simplify the interface using two BOOL properties: `requireCardCVV` and `requireCardPostalCode`.

File: Specs/Braintree-Acceptance-Specs/BTDropInAcceptanceSpec.m [hunk 2]
```diff
@@ -118,6 +141,78 @@
         });
     });
 
+    fdescribe(@"card form", ^{
```
Comment (intelliot) [inline Specs/Braintree-Acceptance-Specs/BTDropInAcceptanceSpec.m]: Probably don't want the `f` here. But when I removed it, I found that other test(s) don't pass

Comment (burnto) [inline Specs/Braintree-Acceptance-Specs/BTDropInAcceptanceSpec.m]: Eek, good catch. Thanks. I'll take a look.

File: Braintree/Drop-In/BTDropInViewController.h [hunk 3]
```diff
@@ -86,13 +93,45 @@
...
+#pragma mark - Card Form
+
+/// The card number.
 ///
-/// @note For the best user experience, you should call this method as early as
-///       possible (after initializing BTDropInViewController, before presenting it)
-///       in order to avoid a loading spinner.
-- (void)fetchPaymentMethods;
+/// If you set a card number longer than is allowed by the card type,
+/// it will not be set.
```
Comment (mickeyreiss) [inline Braintree/Drop-In/BTDropInViewController.h]: This comment is kind of indirect in describing what this property actually does.

File: Braintree/Drop-In/BTDropInViewController.h [hunk 4]
```diff
@@ -86,13 +93,45 @@
...
+@property (nonatomic, copy) NSString *cardNumber;
+
+/// The card form view's expiration date month
```
Comment (mickeyreiss) [inline Braintree/Drop-In/BTDropInViewController.h]: Same with this one. I might suggest talking about default values. `The default expiration month for the card form within Drop in.`

File: Braintree/Drop-In/BTDropInViewController.m [hunk 5]
```diff
@@ -652,6 +652,72 @@ - (void)fetchPaymentMethods {
...
+- (void)setCardCVV:(NSString *)cardCVV {
+    self.dropInContentView.cardForm.cvv = cardCVV;
+}
+
+- (BOOL)requireCardCVV {
+    return self.dropInContentView.cardForm.optionalFields & BTUICardFormOptionalFieldsCvv;
+}
+
+- (void)setRequireCardCVV:(BOOL)requireCardCVV {
+    if (self.requireCardCVV != requireCardCVV) {
```
Comment (mickeyreiss) [inline Braintree/Drop-In/BTDropInViewController.m]: Why worry about this here? Can we just push this check into the card form?

File: Braintree/Drop-In/BTDropInViewController.m [hunk 6]
```diff
@@ -652,6 +652,72 @@ - (void)fetchPaymentMethods {
...
+        self.dropInContentView.cardForm.optionalFields ^= BTUICardFormOptionalFieldsCvv;
```
Comment (mickeyreiss) [inline Braintree/Drop-In/BTDropInViewController.m]: `^=` is slightly tricky...?

File: Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m [hunk 7]
```diff
@@ -39,21 +43,65 @@ - (void)viewDidLoad {
...
-    dropIn.callToActionText = @"$19 - Subscribe Now";
-    dropIn.shouldHideCallToAction = NO;
+- (void)tappedButton {
+
+    if (self.useCardIOSwitch.on) {
+        CardIOPaymentViewController *cardIO = [[CardIOPaymentViewController alloc] initWithPaymentDelegate:self];
+        cardIO.useCardIOLogo = YES;
+        cardIO.suppressScanConfirmation = YES;
+        [self presentViewController:cardIO animated:YES completion:nil];
+    } else{
```
Comment (mickeyreiss) [inline Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m]: whitespace? (`} else {`)

File: Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m [hunk 8]
```diff
@@ -39,21 +43,65 @@ - (void)viewDidLoad {
...
+        BTDropInViewController *dropIn = [self buildDropInViewController];
+        if (self.prepopulateDataSwitch.on) {
+            NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
+            [dateFormatter setDateFormat:@"MM/yyyy"];
+            [dropIn setCardExpirationDate:[dateFormatter dateFromString: @"12/2018"]];
```
Comment (mickeyreiss) [inline Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m]: whitespace

File: Braintree/Drop-In/BTDropInViewController.m (hunk 5 above)
Comment (burnto) [inline Braintree/Drop-In/BTDropInViewController.m]: Well, the card form uses a bitfield, and we're abstracting it here with `BOOL`s.

We could alternatively just expose the bitfield directly, but this seems easier to reason about.

File: Braintree/Drop-In/BTDropInViewController.m (hunk 6 above)
Comment (burnto) [inline Braintree/Drop-In/BTDropInViewController.m]: Open to suggestions, but this seems simple and idiomatic for toggling a bit. :smile:

File: Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m (hunk 7 above)
Comment (burnto) [inline Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m]: :green_apple: willfix.

File: Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m (hunk 8 above)
Comment (burnto) [inline Demos/Braintree-Demo/Features/Drop In/BraintreeDemoDropInViewController.m]: :green_apple: willfix.

**Pull request closed without merging**